Added support for initial-repeat-time #3712

systemmonkey42 opened the pull request:
This patch adds a new global option called `initial-repeat-time`.  This provides the means to differentiate between the time taken for the first key repeat, and the time between subsequent key repeats.

This helps to reduce the instances where a repeating key, once released, triggers
additional repeats when it should be sent to the active pane.

A typical example is using <kbd>PREFIX-ALT_UP</kbd> to resize a pane.  After resizing a few lines, then releasing <kbd>ALT</kbd>, you would normally have to wait a second before pressing  <kbd>UP</kbd> to recall a shell command.  With this patch, the delay is tunable to a much smaller value.

I currently use the following values which work well on remote sessions
(local sessions can often have a smaller "repeat-time")

```
set-option -g repeat-time 350
set-option -g initial-repeat-time 900
```

Patch includes the addtional man page entries.

nicm commented:
This looks good but I think either `initial-repeat-time` should default to zero *or* we should change to your 900/350 suggestion...

nicm commented:
I have made a few changes to your diff. I've rewritten the man page info and option description to be a bit clearer, and also broken the repeat check out into a helper function to avoid some long lines in `server-client.c`. Can you please give this a look/test and make sure I haven't broken it? [repeat.diff.txt](https://github.com/user-attachments/files/17266197/repeat.diff.txt) Thanks

systemmonkey42 commented:
Testing now.

> I think either initial-repeat-time should default to zero

I tend to agree the initial should default to zero, with the repeat at the default value of 500,  Then users who need it can bump `initial-repeat-time` to something higher.  A good combination is having `initial` slightly higher than 500, and the repeat time slighly lower.  600/350 works well on the connection I'm currently testing on.

For me, the default "repeat-time" of "500" is too short to ever get standard key-repeat to kick in.  If I press `PREFIX` and hold `CTRL-UP`, the active pane is resized by one line, then the shell command history starts scrolling.

I suspect the 500 is short is because my initial key repeat is almost exactly 500 ms, however `CTRL-UP` for example is 6 characters, which seems to take almost 100ms to transmit and process.  I suspect if I were directly on a console, I would have a different experience.

systemmonkey42 commented:
I've rebased on the latest build, and implemented your patches.

Seems to work perfectly.  I've reset the repeat-time/initial-repeat-time to 500/0 to see how that works.

nicm commented:
Great, I have applied this now - let me know if you think we should change the defaults. Thanks!
